Enhancing and addressing some issues around exceptions:

toumorokoshi opened the pull request:
This PR attempts to address a few issues around exceptions:

#452: all exceptions are now hashable and immutable
#445: exceptions now have a human readable message, and a to_primitive function is provided to for serialization in markup (e.g. json)
#369: conforms to PEP-352, and normalizes all exceptions to a message and an errors field.

This PR has a backward-incompatible change of FieldErrors no longer being mutable, which I'm amenable to. FieldError.pop() is removed as a result.

toumorokoshi commented:
@lkraider here's a PR to unblock #465

lkraider commented:
Thanks for the PR and addressing the mutability as well, I'll review it today!

lkraider commented on schematics/exceptions.py:
```diff
@@ -2,12 +2,76 @@
+
+    @property
+    def errors(self):
+        return copy.deepcopy(self._errors)
```
**Improvement**: The deepcopy is safe but could be an expensive side-effect for a property to wrap, can't we use an immutable primitive for the errors? A tuple of namedtuples, or something like https://github.com/slezica/python-frozendict ?

toumorokoshi replied:
I'd vote for frozendict, the dict representation is nice without having to serialize to / from tuples. 

Would you like me to pull frozendict as a dependency, or write one for schematics?

lkraider replied:
I think we can pull a copy of frozendict into datastructures.py, it's simple enough.

lkraider commented on schematics/exceptions.py:
```diff
@@ -66,55 +129,55 @@ def __init__(self, *args, **kwargs):
+            msg_repr = '[' + str.join(', ', map(str, self.errors)) + ']'
         return '%s(%s)' % (self.__class__.__name__, msg_repr)
 
     __str__ = __repr__
```
We need to solve the quoting issue and probably return a JSON parseable string:

```python
try:
    Player({'id': 1.2, 'location': [2]})
except DataError as e:
    print(str(e))
```

Now:
```
{'id': u'ConversionError("Value \'1.2\' is not int.")', 'location': u'ConversionError("Input must be a mapping or \'Location\' instance")'}
```

Desired:
```
{"id": "Value '1.2' is not int.", "location": "Input must be a mapping or 'Location' instance"}
```

toumorokoshi replied:
sounds good, I can make the string more legible. 

I feel like it shouldn't be guaranteed behavior that the string cast is json parseable: it's a strange restriction to put on a call that's supposed to be a human-readable representation of the errors.

I'd argue for documenting the use of something like to_primitive() for those who want to serialize output in whatever format they desire (e.g. json or yaml)

lkraider replied:
Indeed, the use case is to be human-readable. I was thinking of the json structure for the simplicity of serializing the dict into a `field_name: message` string, what structure would you propose?

toumorokoshi commented:
Here's an updated version, addressing some of the PR feedback:

1. string representations of error messages are equivalent to json formatting a primitive representation. Agreed json is a pretty legible format, just want to note it's an implementation detail to serialize to json, and shouldn't be relied on. (serializing to_primitive exists for that purpose)

1a. this also involved modifying the string formatting around error message a bit, so it doesn't write out as tuples in the case there's a summary and info included: it was clunky to put a bunch of escaped quotes in the values, if we're targeting human readability.

2. dicts and lists are converted to FrozenDict and FrozenList respectively. to note: I wrote my own implementation of FrozenDict to avoid pulling in the MIT license of the FrozenDict code. modified unit tests that relied on mutable lists.

3. I spent some time examining the FieldError code, and was able to pull more of that checking into the BaseError. This ensures that all the conveniences of FieldError (e.g. __eq__ of the exception with the underying list or dict) is now consistent for all BaseError types.

toumorokoshi commented on schematics/exceptions.py:
```diff
@@ -26,49 +29,71 @@ def __init__(self, message, errors=None, *args):
+        return self._primitive
+
+    @staticmethod
+    def _freeze(obj):
```
I wanted to note that it's still possible to modify elements and dictionary values of FrozenDict and FrozenList. At a performance cost, I can do a recursive _freeze to guard against this.

lkraider replied:
I think we can keep it so, it's pretty hard to enforce complete immutability in Python anyways, and making the base object immutable is a good enough indication (for consenting adults) that it shouldn't be changed.

lkraider commented on schematics/exceptions.py:
```diff
@@ -2,9 +2,98 @@
+    def __repr__(self):
+        return json.dumps(self.to_primitive())
+
+    __str__ = __repr__
```
Do we need `repr` to be equal `str`? We can have repr return a "reconstructable" string, as that is its recommended format.

lkraider commented on tests/test_exceptions.py:
```diff
@@ -86,7 +87,7 @@ def test_error_eq():
 
 
-def test_error_pop():
+def _test_error_pop():
```
Let's remove this test, this functionality will not be supported.

lkraider commented on schematics/exceptions.py:
```diff
@@ -2,9 +2,98 @@
 from __future__ import unicode_literals, absolute_import
 
-from collections import Sequence
+import copy
```
Unused import of `copy`.

toumorokoshi commented:
@lkraider addressing the feedback.

I was able to make repr work with the eval, as specified.

I also noticed something about PEP 352: the only reference to using a string as the first argument is for the special casing that is included in the BaseException class: it looks like the type of the first argument is not so important.

even the examples of user-defined exceptions violates the message as the first argument:

https://docs.python.org/3/tutorial/errors.html

As such, I modified the behavior to not take a string, and to always take a list or dict object as the first argument. This gets rid of the clumsy user-driven string, which was not used anywhere.

bintoro commented:
Once this work is complete, could one of you squash the commits together, please?

toumorokoshi commented:
I would squash, but I believe that erases comment history. I'll leave it up to the merger unless it's ok for me to overwrite history here. 

The repr check has an issue with the unicode encoding and Python2. I'll try to tackle it tomorrow.

lkraider commented:
@toumorokoshi The code is looking great!

@bintoro sure thing, I'll squash&merge from the github interface when it's done.

toumorokoshi commented:
@lkraider so in tackling that repr issue, I found another fun issue. I've resolved it to a pretty good point, but I wanted to make note of it.

the str_compat decorator on classes was interfering with the ability to make nested repr calls. In the case of CompoundError in Python2, the ErrorMessage class would be converted to a unicode-escaped string, which would then be double-escaped when the BaseError escapes ITS __repr__ call.

Removing str_compat form BaseError and ErrorMessage resolves the issue, but I imagine it loses the __str__ and __unicode__ behavior that was expected in the first place. Thus, I split the str_compat into str_compat and repr_compat. I ran into a weird recursive import with attempting to use repr_compat on the model. I put in a quick workaround, but I'll take a closer look if I can. 

Ultimately, the rational for operating on __str__ / __unicode__ makes sense to me, but attempting to escape __repr__ makes nested repr calls impossible, and it's unnecessary as long as __repr__ overrides compose their parts out of other __repr__ calls (such as how BaseError / ErrorMessage now works).

At this point this addresses all the feedback, and maintains as much compatibility as possible (the broken places were already mentioned above, like mutability).

it might be fun to tackle the repr_compat issue across all of schematics in a separate PR. But if desired I can do it in this one.

toumorokoshi commented:
Ok, I think this is good to go now. I removed the hacky with_repr in str_compat (it was an ImportError hidden under a few other messages), and fixed a flakey unit test.

lkraider commented:
@toumorokoshi Thank you for the great patch, squashed and applied to development!

toumorokoshi commented:
Thanks! Great feedback, pleasure to work with you, and looking forward to the 2.0 release!
